### DISCUSSION TITLE
parallel speed up with joblist

### DISCUSSION THREAD

**Discussion by kjmaki** [capytaine/capytaine#447] [Q&A] [answered]
Hello,

I am trying to use the joblist mechanism to speed up calculation of a range of wave frequencies.  I have read through the discussions and issues, and still can't figure out how to realize a speed up.  

Here is code:

```
import logging

import numpy as np
from numpy import pi
import xarray as xr
import matplotlib.pyplot as plt

import capytaine as cpt

logging.basicConfig(level=logging.INFO)

# SET UP THE MESH
buoy = cpt.FloatingBody(
        mesh=cpt.mesh_vertical_cylinder(
            radius=1.0, length=3.0, center=(0, 0, -1.0),
            resolution=(15, 15, 3)
            ))
buoy.add_translation_dof(name="Surge")

# SOLVE THE BEM PROBLEMS AND COMPUTE THE KOCHIN FUNCTIONS
theta_range = np.linspace(0, 2*pi, 40)
omega_range = np.linspace(1.0, 5.0, 3)

test_matrix = xr.Dataset(coords={
    'omega': omega_range, 'theta': theta_range, 'radiating_dof': ["Surge"],
})
solver = cpt.BEMSolver()

from datetime import datetime, timedelta
timeBefore = datetime.now()

dataset = solver.fill_dataset(test_matrix, buoy.immersed_part(),n_jobs=2, wavenumber=True)

timeAfter = datetime.now()
print("total time = ", timeAfter - timeBefore)

# PLOT THE KOCHIN FUNCTION
plt.figure()
dataset = dataset.swap_dims({'omega': 'wavenumber'})  # Use wavenumber to index data
for k in dataset.coords['wavenumber']:
    (dataset['kochin']
     .sel(radiating_dof="Surge", wavenumber=k)
     .real
     .plot(x='theta', label=f"k={float(k):.2f} m¯¹", ax=plt.gca()))
plt.legend()
plt.title("Real part of the Kochin function as a function of theta")
plt.tight_layout()
plt.show()

```
When I change the parameter n_jobs to 2, the computational time actually slows down. Does anyone have any clues on what I am doing wrong?

Thanks!  

Kevin

**Comment by mancellin** [comment, marked as answer]
My guess is that there are only 3 independent problems (you have 3 frequencies in `omega_range`) to be solved in parallel by 2 workers. If you add the small overhead due to the dispatch between workers, it could be overall slower.

Note also that some parallel computation is happening at another level even when `n_jobs=1`. So in some cases, you might not notice a difference when setting `n_jobs=2`, because in both cases all CPU cores will be busy.

**Reply by kjmaki** [reply]
Thanks for your reply.  I tested on two other machines with up to 8 cores.  I can't seem to see any speed up (usually a slow down), but I bet you are right that the parallelization is happening at another level, since it runs faster on the 8 core machine (although hardware is different so I don't want to imply with certainty that it is happening).
